### PR TITLE
[APPWIZ] Creating internet shortcut

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

This PR will add supports of internet shortcuts in appwiz.
JIRA issue: [CORE-8737](https://jira.reactos.org/browse/CORE-8737)

1. Win+R "C:\reactos\system32\rundll32 appwiz.cpl,NewLinkHere C:\"
2. Enter "https://reactos.org".
3. Enter.

It creates "C:\New Internet Shortcut.url".

### CODE REVIEW

**Comment by katahiromz** [comment]
STEP 1:
appwiz-test-1
STEP 2:
appwiz-test-2
STEP 3:
appwiz-test-3

File: dll/cpl/appwiz/createlink.c [hunk 1]
```diff
@@ -141,8 +142,44 @@ CreateShortcut(PCREATE_LINK_CONTEXT pContext)
...
+
+    hr = pPersistFile->lpVtbl->Save(pPersistFile, szPath, TRUE);
+
+    pPersistFile->lpVtbl->Release(pPersistFile);
+    pURL->lpVtbl->Release(pURL);
+
+    return SUCCEEDED(hr);
+}
 
+BOOL IsInternetLocation(LPCWSTR pszLocation)
```
Comment (learn-more) [inline dll/cpl/appwiz/createlink.c]: `PathIsURL` does not work?

File: dll/cpl/appwiz/createlink.c [hunk 2]
```diff
@@ -347,13 +413,18 @@ ShowCreateShortcutWizard(HWND hwndCPl, LPWSTR szPath)
         return FALSE;
     }
 
-    wcscpy(pContext->szLinkName, szPath);
-    if (pContext->szLinkName[nLength-1] != L'\\')
+    /* build the pContext->szOrigin and pContext->szOldFile */
+    wcscpy(pContext->szOrigin, szPath);
```
Comment (learn-more) [inline dll/cpl/appwiz/createlink.c]: `StringCchCopyW` or `wcscpyn` or `wcscpy_s`

Comment (katahiromz) [reply dll/cpl/appwiz/createlink.c]: Used StringCchCopyW.

File: dll/cpl/appwiz/createlink.c (hunk 1 above)
Comment (katahiromz) [reply dll/cpl/appwiz/createlink.c]: OK.

**Comment by katahiromz** [comment]
Done.

**Comment by blackcrack** [comment]
cool, is therewith also possible for create a ftp, sftp, and other shortcuts with login alias:passwd@ ?
If you are in the process of  ;) 
:))) :+1:  love it !

**Comment by katahiromz** [comment]
Any protocols that shlwapi PathIsURL function supported are supported.

**Comment by learn-more** [comment]
no @blackcrack, one feature at a time.

File: dll/cpl/appwiz/appwiz.h [hunk 3]
```diff
@@ -17,6 +17,10 @@
 #include <winreg.h>
 #include <winnls.h>
 #include <shlobj.h>
+#include <intshcut.h>
+#include <shlwapi.h>
+#include <tchar.h>
```
Comment (HBelusca) [inline dll/cpl/appwiz/appwiz.h]: tchar.h should not be needed.

File: dll/cpl/appwiz/createlink.c [hunk 4]
```diff
@@ -299,16 +324,56 @@ FinishDlgProc(HWND hwndDlg,
...
+
+                    /* change extension if any */
+                    pch = PathFindExtensionW(pContext->szLinkName);
+                    if (pch && *pch)
+                        *pch = 0;
+                    StringCchCatW(pContext->szLinkName, _countof(pContext->szLinkName), L".url");
+
+                    if (!CreateInternetShortcut(pContext))
+                    {
+                        MessageBox(hwndDlg, _T("Failed to create internet shortcut"), NULL, MB_ICONERROR);
```
Comment (HBelusca) [inline dll/cpl/appwiz/createlink.c]: MessageBoxW, and use L"..." for the string.

File: dll/cpl/appwiz/createlink.c [hunk 5]
```diff
@@ -299,16 +324,56 @@ FinishDlgProc(HWND hwndDlg,
...
+
+                    /* change extension if any */
+                    pch = PathFindExtensionW(pContext->szLinkName);
+                    if (pch && *pch)
+                        *pch = 0;
+                    StringCchCatW(pContext->szLinkName, _countof(pContext->szLinkName), L".lnk");
+
+                    if (!CreateShortcut(pContext))
+                    {
+                        MessageBox(hwndDlg, _T("Failed to create shortcut"), NULL, MB_ICONERROR);
```
Comment (HBelusca) [inline dll/cpl/appwiz/createlink.c]: Same comment as above.

File: dll/cpl/appwiz/appwiz.h (hunk 3 above)
Comment (katahiromz) [reply dll/cpl/appwiz/appwiz.h]: <strsafe.h> requests <tchar.h> inclusion. It's a necessary.

Comment (HBelusca) [reply dll/cpl/appwiz/appwiz.h]: ah ok then, sorry :)

File: dll/cpl/appwiz/createlink.c (hunk 4 above)
Comment (katahiromz) [reply dll/cpl/appwiz/createlink.c]: OK.

File: dll/cpl/appwiz/createlink.c (hunk 5 above)
Comment (katahiromz) [reply dll/cpl/appwiz/createlink.c]: OK.

**Comment by katahiromz** [comment]
Done.

**Comment by Mouvedia** [comment]
@katahiromz did you handle `IsShortcut` and `NeverShowExt` registry values?
BTW .url files also support `file://`.

**Comment by katahiromz** [comment]
@Mouvedia Unrelated to this PR. It would be another work.

**Comment by katahiromz** [comment]
@ThFabba Done.

File: dll/cpl/appwiz/createlink.c [hunk 6]
```diff
@@ -5,13 +5,14 @@
...
  *                  Dmitry Chapyshev (lentind@yandex.ru)
  *                  Johannes Anderwald
+ *                  Katayama Hirofumi MZ (katayama.hirofumi.mz@gmail.com)
  * UPDATE HISTORY:
  *      06-17-2004  Created
  */
 
 #include "appwiz.h"
-
 #include <tchar.h>
```
Comment (ThFabba) [inline dll/cpl/appwiz/createlink.c]: This is unnecessary and can be removed

Comment (katahiromz) [reply dll/cpl/appwiz/createlink.c]: In GCC build, #include <strsafe.h> requests <tchar.h> inclusion. It's a necessary to build.

Comment (katahiromz) [reply dll/cpl/appwiz/createlink.c]: https://stackoverflow.com/questions/16852925/need-to-include-strsafe-h-after-tchar-h-error

Comment (katahiromz) [reply dll/cpl/appwiz/createlink.c]: Sorry, I misgot. I will fix it.

Comment (katahiromz) [reply dll/cpl/appwiz/createlink.c]: OK. Done.

**Comment by gonzoMD** [comment]
It would be great if you could set the correct icon for URL files too

**Comment by blackcrack** [comment]
@learn-more : <j0k3>hell dammed, so it is only a 8bit system ! Only one step and one completion</j0k3>
 ;-)
(After all, you don't say I'm supposed to go and program it myself *s*)
best regards
Blacky

**Pull request merged**